### ISSUE TITLE
Data sourcing interface

### ISSUE THREAD

**Issue by tz70s** [dsngroup/broke#39]
@allenlee820202 , take a look at [this](https://ci.apache.org/projects/flink/flink-docs-release-1.2/api/java/org/apache/flink/streaming/api/functions/source/SourceFunction.html)

You can implement this interface for data sourcing.

```java
public class LatencyAwareConsumer<T> implements SourceFunction<T> {
     private volatile boolean isRunning = true;

     private BlockClient blockClient;

     // Additional note, this method is driven via consumer, 
     // therefore, the on_message_arrived may not suitable.
     // You have some choices to extend.
     // 1. Add a thread safe queue to kept up receiving message, and collect via this method.
     // ( This may also benefit, b.c. 
     // you can monitor your queue sizes instead of handling external monitoring metrics. )
     // 2. Find out another way.
     public void run(SourceContext<T> ctx) {
        // When message arrived.
        ctx.collect(payload)
     }

     public void cancel() {
         isRunning = false;
     }
     
     public LatencyAwareConsumer(String brokerAddress) {
         this.blockClient = new BlockClient(brokerAddress);
     }
 }

```

And used like this, in [example](https://ci.apache.org/projects/flink/flink-docs-release-1.4/quickstart/setup_quickstart.html#run-the-example)
```java
// get the execution environment
final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

// add customized data sourcing class
DataStream<String> text = env.addSource(new LatencyAwareConsumer<String>("localhost:1883"))
```

You can also take a look at kafka implementation, it has watermark related arguments.

**Comment by allenlee820202** [comment]
OK, thanks. 

I found a mqtt source function example [here](https://github.com/pkhanal/flink-watson-iot-connector/blob/master/src/main/java/io/github/pkhanal/sources/MQTTSource.java). In the [javadoc](https://ci.apache.org/projects/flink/flink-docs-release-1.4/api/java/), it says the ```run()``` method can run for as long as necessary. I think the pull-based message callback is still feasible. The ```run() ``` method can be blocked as in the example and the callback function collects the message as the message arrived. 
```Java
client.subscribe(properties.getProperty(MQTTSource.TOPIC), (topic, message) -> {
            String msg = new String(message.getPayload(), StandardCharsets.UTF_8);
            ctx.collect(msg);
        });
```

**Comment by tz70s** [comment]
Ok, it's fine you choose any approach you like. But this blocking way may not be efficient, in my opinion. Also, you **may** need to depend on external monitoring metrics, as the way we discuss yesterday.

**Comment by tz70s** [comment]
Fixed in #40
